From e90757c1be6a19b9f9839d76211d15b0e0b270ed Mon Sep 17 00:00:00 2001
From: Leandro Santiago <leandrosansilva@gmail.com>
Date: Sun, 17 Dec 2023 21:57:13 +0100
Subject: [PATCH] 'pwr' operator was added/

---
 simpleRPNcalculator.go      | 17 ++++++++++-------
 simpleRPNcalculator_test.go |  2 ++
 2 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/simpleRPNcalculator.go b/simpleRPNcalculator.go
index 4769283..aaeed16 100644
--- a/simpleRPNcalculator.go
+++ b/simpleRPNcalculator.go
@@ -1,6 +1,7 @@
 package main

 import (
+	"math"
 	"strconv"
 	"strings"
 )
@@ -41,15 +42,17 @@ func operate(s *stack, op string) {
 	case "pwr":
 		exponent := pop(s)
 		base := pop(s)
-		res := base
-		if exponent == 0 {
-			res = 1
-		} else {
-			for i := 1.; i < exponent; i++ {
-				res = res * base
-			}
+		res := 1.
+		absExp := int(math.Abs(exponent))
+		for i := 0; i < absExp; i++ {
+			res = res * base
+		}
+		if exponent < 0 {
+			res = 1. / res
 		}
+
 		push(s, res)
+
 	}
 }

diff --git a/simpleRPNcalculator_test.go b/simpleRPNcalculator_test.go
index 4ce597e..4cff936 100644
--- a/simpleRPNcalculator_test.go
+++ b/simpleRPNcalculator_test.go
@@ -64,6 +64,8 @@ func TestCalculator(t *testing.T) {
 	require.Equal(t, 5., calculator("2 3 4 drop +"))
 	require.Equal(t, 66., calculator("2 4 3 pwr +"))
 	require.Equal(t, -1., calculator("3 0 pwr neg"))
+	require.Equal(t, (1. / 64), calculator("4 -3 pwr"))
+
 }

 func TestOperate(t *testing.T) {
--
2.42.1
